src/main/scala/com/github/codelionx/dodo/discovery/CandidateGenerator.scala

```diff
@@ -0,0 +1,17 @@
+package com.github.codelionx.dodo.discovery
+
+import scala.collection.immutable.Queue
+
+
+trait CandidateGenerator {
+  def generateFirstCandidates(columns: Set[Int]): Queue[(List[Int], List[Int])] = {
+    val fixedOrderCols = columns.toList
+    val candidates = fixedOrderCols.flatMap(col1 => fixedOrderCols.drop(fixedOrderCols.indexOf(col1) + 1).map(col2 => (List(col1), List(col2))))
```

Comment:
This could also use `.combinations`:
```scala
val candidates = columns.toSeq
           .combinations(2)
           .map(l => l(0) -> l(1)) // list to tuple: only safe in this special case (we know length == 2)
           .toSeq
Queue(candidates: _*)
```